File: streamlit_interface/pipeline/nlp_filters.py
"""
nlp_filters.py
──────────────
Local spaCy-based filter extractor for Romanian real-estate prompts.
Replaces the remote Ollama parse_vibe round-trip with instant local NLP.

Usage:
    from nlp_filters import extract_filters, get_olx_keywords

    filters = extract_filters("Caut 2 camere langa metrou, pet-friendly")
    # → {"ROOM_COUNT": "2", "HAS_METRO": True, "PET_FRIENDLY": True}

    slugs = get_olx_keywords(filters)
    # → ["metrou", "pet-friendly"]
    # Append to OLX URL: /q-metrou-pet-friendly/
"""

import spacy

nlp = spacy.load("ro_core_news_sm")

# ── Taxonomy: Romanian lemma → internal filter key ───────────────────────
REAL_ESTATE_TAXONOMY = {
    "metrou":              "HAS_METRO",
    "renovat":             "CONDITION_RENOVATED",
    "modern":              "STYLE_MODERN",
    "luminos":             "FEAT_BRIGHT",
    "centrală":            "HAS_HEATING_UNIT",
    "centrala":            "HAS_HEATING_UNIT",
    "parcare":             "HAS_PARKING",
    "garaj":               "HAS_PARKING",
    "balcon":              "HAS_BALCONY",
    "pet-friendly":        "PET_FRIENDLY",
    "animal":              "PET_FRIENDLY",   # "animal de companie"
    "mobilat":             "FURNISHED",
    "nemobilat":           "UNFURNISHED",
    "liniștit":            "FEAT_QUIET",

}

# ── OLX URL slug mapping: filter key → keyword appended to /q-... path ───
FILTER_TO_OLX_SLUG = {
    "HAS_METRO":           "metrou",
    "HAS_PARKING":         "parcare",
    "PET_FRIENDLY":        "pet-friendly",
    "HAS_HEATING_UNIT":    "centrala",
    "HAS_BALCONY":         "balcon",
    "CONDITION_RENOVATED": "renovat",
    "STYLE_MODERN":        "modern",
    "FURNISHED":           "mobilat",
    "FEAT_QUIET":          "linistit",
}


# ── Description-level exclusion phrases (Romanian) ───────────────────────
# If a listing description contains any of these, it explicitly denies the feature.
EXCLUSION_PATTERNS = {
    "PET_FRIENDLY": [
        "nu acceptam animale", "nu acceptăm animale",
        "nu se accepta animale", "nu se acceptă animale",
        "nu se admit animale", "nu accepta animale",
        "fara animale", "fără animale",
        "animale nepermise", "animale nu sunt permise",
        "nu este pet-friendly", "nu e pet-friendly",
        "nu este pet friendly", "nu e pet friendly",
        "no pets", "nu acceptam caini", "nu acceptăm câini",
        "nu acceptam pisici", "nu acceptăm pisici",
    ],
    "HAS_PARKING": [
        "fara loc de parcare", "fără loc de parcare",
        "nu include parcare", "nu este inclusa parcarea",
        "parcare neinclusa", "parcare neinclus",
        "nu are parcare", "nu dispune de parcare",
    ],
    "HAS_BALCONY": [
        "fara balcon", "fără balcon",
        "nu are balcon", "nu dispune de balcon",
    ],
    "FURNISHED": [
        "nemobilat", "fara mobila", "fără mobilă",
        "nu este mobilat", "nu e mobilat",
    ],
    "HAS_HEATING_UNIT": [
        "fara centrala", "fără centrală",
        "nu are centrala", "nu are centrală",
        "incalzire centrala de bloc",   # district heating ≠ own boiler
    ],
}

# ── Positive confirmation phrases ─────────────────────────────────────────
# If present in description, listing explicitly confirms the feature.
CONFIRMATION_PATTERNS = {
    "PET_FRIENDLY": [
        "pet-friendly", "pet friendly", "acceptam animale",
        "acceptăm animale", "animale acceptate", "animale de companie permise",
        "acceptam catei", "acceptăm câini", "acceptam pisici",
    ],
    "HAS_PARKING": [
        "loc de parcare", "parcare inclusa", "parcare inclusă",
        "parcare proprie", "garaj inclus",
    ],
    "HAS_BALCONY": [
        "balcon", "terasa", "terasă",
    ],
    "FURNISHED": [
        "mobilat", "mobilată", "complet mobilat", "partial mobilat",
    ],
}


def apply_description_filters(df, spacy_filters: dict):
    """
    Hard-exclude listings whose descriptions explicitly deny a requested feature.
    Returns (filtered_df, excluded_count, summary_dict).

    summary_dict maps filter_key → count of listings excluded by that filter.
    """
    import pandas as pd

    if not spacy_filters or "description" not in df.columns:
        return df, 0, {}

    desc = df["description"].fillna("").str.lower()
    keep = pd.Series(True, index=df.index)
    summary = {}

    for filter_key in spacy_filters:
        patterns = EXCLUSION_PATTERNS.get(filter_key, [])
        if not patterns:
            continue
        filter_mask = pd.Series(False, index=df.index)
        for pattern in patterns:
            filter_mask |= desc.str.contains(pattern, na=False, regex=False)
        excluded = filter_mask.sum()
        if excluded:
            summary[filter_key] = int(excluded)
        keep &= ~filter_mask

    excluded_total = int((~keep).sum())
    return df[keep].reset_index(drop=True), excluded_total, summary


def extract_filters(text: str) -> dict:
    """
    Run spaCy NLP on the user prompt and return a dict of detected filters.

    Returns:
        e.g. {"ROOM_COUNT": "2", "LOCATION_SECTOR": "4", "HAS_METRO": True,
              "PROPERTY_TYPE": "Garsonieră", "AREA_MIN": 50}
    """
    import re as _re
    doc = nlp(text.lower())
    found = {}
    lowered = text.lower()

    # 1. Room count — numeric token whose syntactic head lemma is "cameră"
    for token in doc:
        if token.like_num and token.head.lemma_ in ("cameră", "camera", "camere"):
            found["ROOM_COUNT"] = token.text
            break

    # 2. Sector number — "sectorul" / "sector" followed by a digit
    for i, token in enumerate(doc):
        if token.lemma_ in ("sector", "sectorul") and i + 1 < len(doc):
            nxt = doc[i + 1]
            if nxt.like_num:
                found["LOCATION_SECTOR"] = nxt.text
                break

    # 3. Keyword taxonomy via lemmatization
    for token in doc:
        lemma = token.lemma_
        if lemma in REAL_ESTATE_TAXONOMY:
            key = REAL_ESTATE_TAXONOMY[lemma]
            if key not in found:
                found[key] = True

    # 4. Exact substring check for multi-word phrases
    multi_word = {
        "pet-friendly":        "PET_FRIENDLY",
        "animal de companie":  "PET_FRIENDLY",
        "lângă metrou":        "HAS_METRO",
        "aproape de metrou":   "HAS_METRO",
        "loc de parcare":      "HAS_PARKING",
    }
    for phrase, key in multi_word.items():
        if phrase in lowered and key not in found:
            found[key] = True

    # 5. Property type — keyword matching
    if "PROPERTY_TYPE" not in found:
        if any(w in lowered for w in ("garsonier", "studio", "garconiera", "garçonieră")):
            found["PROPERTY_TYPE"] = "Garsonieră"
        elif any(w in lowered for w in ("casă", "casa", "vilă", "vila", "duplex")):
            found["PROPERTY_TYPE"] = "Casă / Vilă"
        elif "apartament" in lowered:
            found["PROPERTY_TYPE"] = "Apartament"

    # 6. Area range — "X mp", "minim X mp", "pana la X mp", etc.
    # Minimum area: "minim/cel putin/minimum X mp"
    m = _re.search(
        r"(?:minim|cel\s+pu[tț]in|minimum|peste|mai\s+mare\s+de)\s+(\d+)\s*mp",
        lowered,
    )
    if m:
        found["AREA_MIN"] = int(m.group(1))

    # Maximum area: "maxim/pana la/cel mult X mp"
    m = _re.search(
        r"(?:maxim|p[aâ]n[aă]\s+la|cel\s+mult|sub)\s+(\d+)\s*mp",
        lowered,
    )
    if m:
        found["AREA_MAX"] = int(m.group(1))

    # Bare "X mp" with no qualifier → treat as minimum hint
    if "AREA_MIN" not in found and "AREA_MAX" not in found:
        m = _re.search(r"(\d{2,3})\s*mp", lowered)
        if m:
            found["AREA_MIN"] = int(m.group(1))

    # 7. Price — "maxim 1500 euro", "sub 2000 eur", "pana la 1800 €", "buget 1200", bare "1500 euro"
    _currency = r"(?:euro|eur|€|ron|lei)"
    # Explicit upper-bound phrases
    m = _re.search(
        r"(?:maxim|p[aâ]n[aă]\s+la|cel\s+mult|sub|mai\s+pu[tț]in\s+de|buget(?:\s+de)?|pret\s+maxim)\s*(\d{3,5})\s*" + _currency,
        lowered,
    )
    if m:
        found["PRICE_MAX"] = int(m.group(1))
    else:
        # Bare "1500 euro" / "1500 €" with no qualifier
        m = _re.search(r"(\d{3,5})\s*" + _currency, lowered)
        if m:
            found["PRICE_MAX"] = int(m.group(1))

    return found


def get_olx_keywords(filters: dict) -> list[str]:
    """
    Convert a filters dict into a list of OLX URL slug keywords.

    Example:
        {"HAS_METRO": True, "PET_FRIENDLY": True} → ["metrou", "pet-friendly"]

    These are joined with "-" and inserted into the OLX URL path:
        /q-metrou-pet-friendly/
    """
    return [
        FILTER_TO_OLX_SLUG[key]
        for key in filters
        if key in FILTER_TO_OLX_SLUG and filters[key]
    ]


# ── Quick self-test ───────────────────────────────────────────────────────
if __name__ == "__main__":
    test_prompt = (
        "Caut apartament cu 2 camere in sectorul 4, aproape de metrou. "
        "Vreau sa fie ceva modern si renovat. Pet-friendly si cu parcare."
    )
    f = extract_filters(test_prompt)
    print("Filters :", f)
    print("OLX slugs:", get_olx_keywords(f))

File: mock_ui/lib/nlpFilters.ts
// Lightweight regex-based Romanian filter extractor — a TS port of the spaCy-based
// pipeline in streamlit_interface/pipeline/nlp_filters.py, scoped to what the mock
// UI needs to demonstrate: pulling structured filters out of the free-text "vibe" prompt.

import type { PropertyType, RoomCount, VibeFilterKey, VibeFilters } from "./types";

function stripDiacritics(text: string): string {
  return text
    .replace(/ă/g, "a")
    .replace(/â/g, "a")
    .replace(/î/g, "i")
    .replace(/ș/g, "s")
    .replace(/ş/g, "s")
    .replace(/ț/g, "t")
    .replace(/ţ/g, "t");
}

const BOOLEAN_TAXONOMY: [RegExp, VibeFilterKey][] = [
  [/\bmetrou\b/, "HAS_METRO"],
  [/\brenovat[aă]?\b/, "CONDITION_RENOVATED"],
  [/\bmodern[aă]?\b/, "STYLE_MODERN"],
  [/\bluminos[aă]?\b/, "FEAT_BRIGHT"],
  [/\bcentral[aă]\b/, "HAS_HEATING_UNIT"],
  [/\b(parcare|garaj)\b/, "HAS_PARKING"],
  [/\bbalcon\b/, "HAS_BALCONY"],
  [/\bpet[\s-]?friendly\b/, "PET_FRIENDLY"],
  [/\banimal(e)?\s+de\s+companie\b/, "PET_FRIENDLY"],
  [/\bmobilat[aă]?\b/, "FURNISHED"],
  [/\blinistit[aă]?\b/, "FEAT_QUIET"],
];

const PROPERTY_TYPE_PATTERNS: [RegExp, PropertyType][] = [
  [/\bgarsonier[aă]|garconiera\b/, "Garsoniera"],
  [/\bstudio\b/, "Studio"],
  [/\b(casa|vila|duplex)\b/, "Casa/Vila"],
  [/\bapartament\b/, "Apartament"],
];

export function extractVibeFilters(rawText: string): VibeFilters {
  const found: VibeFilters = {};
  if (!rawText.trim()) return found;

  const normalized = stripDiacritics(rawText.toLowerCase());

  // Room count — "2 camere" / "2 camera"
  const roomMatch = normalized.match(/(\d+)\s*camer[ae]/);
  if (roomMatch) {
    const n = parseInt(roomMatch[1], 10);
    found.ROOM_COUNT = n >= 5 ? "5+" : String(n);
  }

  // Boolean taxonomy
  for (const [pattern, key] of BOOLEAN_TAXONOMY) {
    if (!found[key] && pattern.test(normalized)) {
      found[key] = true;
    }
  }

  // "nemobilat" should not also register FURNISHED
  if (/\bnemobilat[aă]?\b/.test(normalized)) {
    delete found.FURNISHED;
  }

  // Property type
  for (const [pattern, type] of PROPERTY_TYPE_PATTERNS) {
    if (pattern.test(normalized)) {
      found.PROPERTY_TYPE = type;
      break;
    }
  }

  // Area — "minim 50 mp" / "pana la 70 mp" / bare "60 mp"
  const areaMin = normalized.match(/(?:minim|cel putin|minimum|peste|mai mare de)\s+(\d+)\s*mp/);
  if (areaMin) found.AREA_MIN = parseInt(areaMin[1], 10);
  const areaMax = normalized.match(/(?:maxim|pana la|cel mult|sub)\s+(\d+)\s*mp/);
  if (areaMax) found.AREA_MAX = parseInt(areaMax[1], 10);
  if (!found.AREA_MIN && !found.AREA_MAX) {
    const bareArea = normalized.match(/(\d{2,3})\s*mp/);
    if (bareArea) found.AREA_MIN = parseInt(bareArea[1], 10);
  }

  // Price — "maxim 1500 euro" / bare "1500 euro"
  const currency = "(?:euro|eur|ron|lei)";
  const priceExplicit = normalized.match(
    new RegExp(`(?:maxim|pana la|cel mult|sub|mai putin de|buget(?: de)?|pret maxim)\\s*(\\d{3,5})\\s*${currency}`)
  );
  if (priceExplicit) {
    found.PRICE_MAX = parseInt(priceExplicit[1], 10);
  } else {
    const priceBare = normalized.match(new RegExp(`(\\d{3,5})\\s*${currency}`));
    if (priceBare) found.PRICE_MAX = parseInt(priceBare[1], 10);
  }

  return found;
}

// ── Description-level confirmation / exclusion (mirrors apply_description_filters) ──
const EXCLUSION_PATTERNS: Partial<Record<VibeFilterKey, RegExp[]>> = {
  PET_FRIENDLY: [/nu accept[aă]m animale/, /fara animale/, /animale nepermise/, /no pets/],
  HAS_PARKING: [/fara loc de parcare/, /nu (are|include|dispune de) parcare/],
  HAS_BALCONY: [/fara balcon/, /nu are balcon/],
  FURNISHED: [/nemobilat/, /fara mobila/, /nu este mobilat/],
  HAS_HEATING_UNIT: [/fara centrala/, /nu are centrala/],
};

export function isExcludedByDescription(description: string, filters: VibeFilters): VibeFilterKey[] {
  const normalized = stripDiacritics(description.toLowerCase());
  const violated: VibeFilterKey[] = [];
  (Object.keys(filters) as VibeFilterKey[]).forEach((key) => {
    const patterns = EXCLUSION_PATTERNS[key];
    if (!patterns) return;
    if (patterns.some((p) => p.test(normalized))) violated.push(key);
  });
  return violated;
}

// ── End-user-facing copy for each filter key — never expose raw internal keys ──
export const VIBE_FILTER_LABELS: Record<VibeFilterKey, string> = {
  ROOM_COUNT: "camere",
  PROPERTY_TYPE: "tip locuință",
  PRICE_MAX: "preț maxim",
  AREA_MIN: "suprafață minimă",
  AREA_MAX: "suprafață maximă",
  HAS_METRO: "aproape de metrou",
  HAS_PARKING: "parcare inclusă",
  HAS_BALCONY: "cu balcon",
  PET_FRIENDLY: "acceptă animale",
  FURNISHED: "mobilat",
  HAS_HEATING_UNIT: "centrală proprie",
  CONDITION_RENOVATED: "recent renovat",
  STYLE_MODERN: "stil modern",
  FEAT_BRIGHT: "luminos",
  FEAT_QUIET: "zonă liniștită",
};

export function formatVibeFilterValue(key: VibeFilterKey, value: string | number | boolean): string {
  if (typeof value === "boolean") return VIBE_FILTER_LABELS[key];
  if (key === "ROOM_COUNT") return `${value} camere`;
  if (key === "PRICE_MAX") return `preț ≤ ${value} €`;
  if (key === "AREA_MIN") return `suprafață ≥ ${value} m²`;
  if (key === "AREA_MAX") return `suprafață ≤ ${value} m²`;
  return String(value);
}
